cookbooks/prescribed_velocity_ascii_data/prescribed_velocity_ascii_data.cc

```diff
@@ -0,0 +1,350 @@
+          {
+            const std::string filename = prescribed_velocity_field_params_2d.data_directory + prescribed_velocity_field_params_2d.data_file_name;
+            prescribed_velocity_field_data_2d.load_file(filename, simulator_access.get_mpi_communicator());
+
+            std::cout << "Loaded 2D prescribed velocity ascii file" << std::endl;
+          }
+        else
+          {
+            const std::string filename = prescribed_velocity_field_params_3d.data_directory + prescribed_velocity_field_params_3d.data_file_name;
+            prescribed_velocity_field_data_3d.load_file(filename, simulator_access.get_mpi_communicator());
+
+            std::cout << "Loaded 3D prescribed velocity ascii file" << std::endl;
+          }
+        prescribed_velocity_ascii_file_loaded = true;
+      }
+
+    if (prescribe_internal_velocities)
+      {
+        const std::vector< Point<dim> > points = get_unit_support_points_for_velocity(simulator_access);
+        const Quadrature<dim> quadrature (points);
+        FEValues<dim> fe_values (simulator_access.get_fe(), quadrature, update_quadrature_points);
+        typename DoFHandler<dim>::active_cell_iterator cell;
+
+        // Loop over all cells
+        for (cell = simulator_access.get_dof_handler().begin_active();
+             cell != simulator_access.get_dof_handler().end();
+             ++cell)
+          if (! cell->is_artificial())
+            {
+              fe_values.reinit (cell);
+              std::vector<types::global_dof_index> local_dof_indices(simulator_access.get_fe().dofs_per_cell);
+              cell->get_dof_indices (local_dof_indices);
+
+              for (unsigned int q=0; q<quadrature.size(); q++)
+                // If it's okay to constrain this DOF
+                if (current_constraints.can_store_line(local_dof_indices[q]) &&
+                    !current_constraints.is_constrained(local_dof_indices[q]))
+                  {
+                    // Get the velocity component index
+                    const unsigned int c_idx =
+                      simulator_access.get_fe().system_to_component_index(q).first;
+
+                    // If we're on one of the velocity DOFs
+                    if ((c_idx >=
+                         simulator_access.introspection().component_indices.velocities[0])
+                        &&
+                        (c_idx <=
+                         simulator_access.introspection().component_indices.velocities[dim-1]))
+                      {
+                        // Which velocity component is this DOF associated with?
+                        const unsigned int component_direction
+                          = (c_idx
+                             - simulator_access.introspection().component_indices.velocities[0]);
+
+                        // we get time passed as seconds (always) but may want
+                        // to reinterpret it in years
+                        double time = simulator_access.get_time();
```

Comment:
this line and the following several should probably be moved out of the loop over all cells